---
title: "crICSKAT Tutorial"
author: "Zhichao Xu, Jaihee Choi, and Ryan Sun"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{crICSKAT Tutorial}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

## Introduction 
The crICSKAT package is designed to implement the crICSKAT (Interval-Censored Sequence Kernel Association Test accounting for competing risks) and the interval-censored Burden test while accounting for competing risks. This package is particularly useful for making inferences on sets of features in genetic association studies, such as all SNPs in a gene or pathway.

Interval-censored data is a category of failure time data, also known as time-to-event data, that is frequently encountered in contemporary genetic datasets such as the UK Biobank (UKB) and St. Jude Lifetime Cohort Study (SJLIFE). While many individuals are familiar with right-censored data, which is a type of interval-censored data, interval-censored data is not precisely known but is only known to fall within a specific range or interval (such as between 0 and 10, between 15 and 18, or between 30 and infinity).

The follow-up information for individuals in the UK Biobank (UKB) was gathered using questionnaires and sample assays conducted during visit assessments, which were subsequently linked to their medical records. Consequently, providing longitudinal data with time-stamped phenome-wide diagnosis information presents a challenge. Therefore, for some diseases, the onset can only be determined to lie within an interval between two assessments or before the first assessment, rather than being precisely observed, resulting in interval-censored failure times.

Competing risks occur when an individual may experience multiple events that could lead to failure, and these events may compete with each other to cause the failure. This means that the occurrence of one event may affect or impede the likelihood of another event happening. For instance, during the study period, some individuals may have passed away due to other causes, rendering them unable to experience the fracture event or survive until the end of the study period.

There exist some ad-hoc approaches for handling interval-censored data, such as converting the occurrence of the event into a binary outcome and utilizing binary outcome techniques (e.g., logistic regression), or estimating the occurrence time to fall in the middle of the interval and using right-censored survival analysis methods. Nevertheless, when competing outcomes are present, utilizing interval-censored methodology that accounts for these competing risks frequently results in improved operating characteristics, such as better control of the Type I error rate or increased power.

## Steps

To run the crICSKAT test using this package, follow these three steps:

1. Generate the design matrices for the null model in the expected format using the crICSKAT::makeICdmat() function. This step only needs to be performed once for all SNP-sets that will be tested. The function takes the following arguments:
    - xMat is the n*p matrix of non-genetic covariates (not including an intercept)
    - lt is the n*1 vector of times for the left side of the interval. If an observation is left-censored, you can just put in 0.
    - rt is the n*1 vector of times for the right side of the interval. If an observation is right-censored, you can put in Inf or any numeric value.
    - obsInd is the n*1 vector of whether the event was observed before last follow-up (whether it is right-censored). If it was observed before last follow-up (not right-censored), there should be a 0, otherwise there should be a 1.
    - quant_r can be used to pass in knot locations for the spline, we suggest not specifying this and letting the package work automatically.
    - nKnots is a scalar number of interior knots. The default of 1 will create three total knots, one interior and two endpoint knots.

2. Next you need to fit the null model. This only needs to be done once for all sets of SNPs to be tested. To do this, use the crICSKAT::crICSKAT_fit_null() function and provide the following arguments:
    - init_beta is a vector containing the initial guesses for the covariates in each column of the design matrices generated by crICSKAT::makeICdmat(). A vector of zeros or ones can be used for initialization. Providing a good estimate of the coefficients can speed up the convergence process.
    - leftDmat is output directly from crICSKAT::makeICdmat().
    - rightDmat is output directly from crICSKAT::makeICdmat().
    - deltaVec is the n*1 vector of the event cause (cause 1 or 2; or right-censored 0).
    - leftTimes is the same as lt in crICSKAT::makeICdmat().
    - gSummed is the n*1 vector of summed genotype matrix.
    - allowSingular is a logical value indicating if a small correction to the Jacobian when it is singular or too ill-conditioned is allowed.
    - method is the method to use for finding a solution.

3. After fitting the null model, the final step is to test each set of SNPs using the crICSKAT::crICSKAT() function. The gMat argument should be the n*q matrix of genotypes. You will also need leftDmat, rightDmat, leftTimes, gSummed, and deltaVec, which have been covered above. Finally, you will need null_beta, which come directly from crICSKAT::crICSKAT_fit_null(), see below for an example.

## Worked Example 

Suppose we aim to investigate the potential association between a particular gene and time to bone mineral density deficiency. To explore this hypothesis, we will simulate event times for 5,000 subjects using a proportional hazards model with a baseline cumulative hazard function H(t)=t. We will designate five observation times at times 1, 2, 3, 4, and 5, with each subject's exact visit times randomly drawn from a Uniform(-0.25, 0.25) distribution surrounding these times. Moreover, each subject has a 10% likelihood of missing any given visit.

The genetic data will encompass 50 single nucleotide polymorphisms (SNPs) within the gene of interest. For each patient, we have recorded their minor allele count (0,1,2) at each of these 50 SNPs. In addition, we have information on non-genetic covariates for each subject's gender (binary) and their weight (continuous).

```{r}
library(ICSKAT)
set.seed(0)
n <- 5000
q <- 50
# generate data
# all SNPs have minor allele frequency 0.3 in this toy example
gMat <- matrix(data=rbinom(n=n*q, size=2, prob=0.3), nrow=n)
# gender and whether they take daily vitamins
xMat <- matrix(data=rbinom(n=n*2, size=1, prob=0.5), nrow=n)
# the baseline cumulative hazard function
bhFunInv <- function(x) {x}
# observation times
obsTimes <- 1:4
# no effect of either gender or daily vitamins
etaVec <- rep(0, n)
# generate data
outcomeDat <- gen_IC_data(bhFunInv = bhFunInv, obsTimes = obsTimes, windowHalf = 0.25,
                          probMiss = 0.1, etaVec = etaVec)
lt <- outcomeDat$leftTimes
rt <- outcomeDat$rightTimes
# indicators of left- and right-censoring
tpos_ind <- as.numeric(lt > 0)
obs_ind <- as.numeric(rt != Inf)
# make design matrix with cubic spline terms
dmats <- make_IC_dmat(xMat, lt, rt, obs_ind, tpos_ind)
# fit null model - only need to do this once for each genetic set (note there is no information
# on the SNPs used here)
nullFit <- ICSKAT_fit_null(init_beta = rep(0, 5), left_dmat = dmats$left_dmat, right_dmat=dmats$right_dmat, 
                           obs_ind = obs_ind, tpos_ind = tpos_ind, lt = lt, rt = rt)
# perform the ICSKAT and Burden tests
icskatOut <- ICskat(left_dmat = dmats$left_dmat, right_dmat=dmats$right_dmat, lt = lt, rt = rt,
       obs_ind = obs_ind, tpos_ind = tpos_ind, gMat = gMat, null_beta = as.numeric(nullFit$beta_fit), Itt = nullFit$Itt)
icskatOut$p_SKAT
icskatOut$p_burden
# perform the ICSKATO test
ICSKATO(icskatOut = icskatOut)$pval
```

If we want to test another SNP set (e.g. another gene), we don't need to run the make_IC_dmat or ICSKAT_fit_null functions again. Just run ICskat() on the new genotype matrix. 

```{r}
# another gene with 100 SNPs in it
gMat2 <- matrix(data=rbinom(n=n*100, size=2, prob=0.3), nrow=n)
# we don't need to run the make_IC_dmat or ICSKAT_fit_null functions again as long
# as the event times and non-genetic covariates haven't changed.
icskatOut2 <- ICskat(left_dmat = dmats$left_dmat, right_dmat=dmats$right_dmat, lt = lt, rt = rt,
       obs_ind = obs_ind, tpos_ind = tpos_ind, gMat = gMat2, null_beta = as.numeric(nullFit$beta_fit), Itt = nullFit$Itt)
icskatOut2$p_SKAT
icskatOut2$p_burden
# perform the ICSKATO test
ICSKATO(icskatOut = icskatOut2)$pval
```

Questions or novel applications? Please let me know!  Contact information can be found in the package description.



